internal/internal_eager_activity.go

```diff
@@ -0,0 +1,157 @@
+)
+
+// eagerActivityExecutor is a worker-scoped executor for eager activities that
+// are returned from workflow task completion responses.
+type eagerActivityExecutor struct {
+	eagerActivityExecutorOptions
+
+	activityWorker *activityWorker
+	pendingCount   int // Only access under countLock
+	executingCount int // Only access under countLock
+	countLock      sync.Mutex
+}
+
+type eagerActivityExecutorOptions struct {
+	disabled  bool
+	taskQueue string
+	// If 0, there is no maximum
+	maxConcurrent int
+}
+
+// newEagerActivityExecutor creates a new worker-scoped executor without an
+// activityWorker set. The activityWorker must be set on the responding executor
+// before it will be able to execute activities.
+func newEagerActivityExecutor(options eagerActivityExecutorOptions) *eagerActivityExecutor {
+	return &eagerActivityExecutor{eagerActivityExecutorOptions: options}
+}
+
+func (e *eagerActivityExecutor) applyToRequest(
+	req *workflowservice.RespondWorkflowTaskCompletedRequest,
+) (amountReserved int) {
+	// Go over every command checking for activities that can be eagerly executed
+	for _, command := range req.Commands {
+		if attrs := command.GetScheduleActivityTaskCommandAttributes(); attrs != nil {
+			// If not present, disabled, or on a different task queue, we must mark as
+			// explicitly disabled
+			if e == nil || e.disabled || e.activityWorker == nil || e.taskQueue != attrs.TaskQueue.GetName() {
+				attrs.RequestEagerExecution = false
+			} else if attrs.RequestEagerExecution {
+				// If it has been requested, attempt to reserve one pending
+				attrs.RequestEagerExecution = e.reserveOnePendingSlot()
+				if attrs.RequestEagerExecution {
+					amountReserved++
+				}
+			}
+		}
+	}
+	return
+}
+
+func (e *eagerActivityExecutor) reserveOnePendingSlot() bool {
+	// Lock during count checks. Nothing in here blocks including the channel
+	// receive to serve a slot.
+	e.countLock.Lock()
+	defer e.countLock.Unlock()
+	// Confirm that, if we have a max, pending + executing isn't already there
+	if e.maxConcurrent > 0 && e.executingCount+e.pendingCount >= e.maxConcurrent {
+		// No more room
+		return false
+	}
+	// Reserve a spot for our request via a non-blocking attempt to take a poller
+	// request entry which essentially reserves a spot
+	select {
+	case <-e.activityWorker.worker.pollerRequestCh:
+	default:
+		return false
+	}
+
+	// We can request, so increase the pending count
+	e.pendingCount++
+	return true
+}
+
+func (e *eagerActivityExecutor) handleResponse(
+	resp *workflowservice.RespondWorkflowTaskCompletedResponse,
+	amountReserved int,
+) {
+	// Ignore disabled or none present
+	if e == nil || e.activityWorker == nil || e.disabled || (len(resp.ActivityTasks) == 0 && amountReserved == 0) {
+		return
+	} else if len(resp.ActivityTasks) > amountReserved {
+		panic(fmt.Sprintf("Unexpectedly received %v eager activities though we only requested %v",
+			len(resp.ActivityTasks), amountReserved))
```

Comment:
I don't think panic helps much here. Definitely a server bug, and worth printing an error, but in Core I print the error and then drop them on the floor. They'll run into their timeout eventually. This will result in the same thing, but also the worker crashes.

Reply:
Should probably also print errors if the response is trying to eagerly execute stuff but any of this part of the clause is true:

`e == nil || e.activityWorker == nil || e.disabled`

Reply:
I'd actually argue that it's better to panic and that we make sure we run integration tests with pre-release server versions to avoid server bugs popping up.

Reply:
Yeah, this is basically an assertion that server doesn't do something we didn't ask it too. I think it's reasonable to fail like a never-should-fail assertion.

Reply:
Panic only during test would make a lot of sense too, I do that in a handful of places

Reply:
Whether you are testing is not a knowable thing in Go (they don't have an `assert` like many other languages). Maybe I'll just log and ignore and give the slots back.

Reply:
I vote for keeping the panic just so we have visibility for bad server behavior

Reply:
Left it there for now